src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -1272,6 +1275,73 @@ export class SavedObjectsRepository {
     }
   }
 
+  async addToWorkspaces(
+    objects: SavedObjectsShareObjects[],
+    workspaces: string[],
+    options: SavedObjectsAddToWorkspacesOptions = {}
+  ): Promise<SavedObjectsAddToWorkspacesResponse[]> {
+    objects.forEach(({ type, id }) => {
+      if (!this._allowedTypes.includes(type)) {
+        throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+      }
+    });
+
+    if (!workspaces.length) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        'workspaces must be a non-empty array of strings'
+      );
+    }
+
+    const { version, refresh = DEFAULT_REFRESH_SETTING } = options;
+    // we do not need to normalize the namespace to its ID format, since it will be converted to a namespace string before being used
+
+    const savedObjectsBulkResponse = await this.bulkGet(objects);
+
+    const docs = savedObjectsBulkResponse.saved_objects.map((obj) => {
+      const { type, id } = obj;
+      const rawId = this._serializer.generateRawId(undefined, type, id);
+      const existingWorkspace = obj.workspaces || [];
+      // there should never be a case where a multi-namespace object does not have any existing namespaces
+      // however, it is a possibility if someone manually modifies the document in OpenSearch
+      const time = this._getCurrentTime();
+
+      return [
+        {
+          update: {
+            _id: rawId,
+            _index: this.getIndexForType(type),
+            ...getExpectedVersionProperties(version),
+          },
+        },
+        {
+          doc: {
+            updated_at: time,
+            workspaces: unique(existingWorkspace.concat(workspaces)),
+          },
+        },
+      ];
+    });
+
+    const bulkUpdateResponse = await this.client.bulk({
```

Comment:
I am thinking if we can use `update_script` to do such `append` logic.

```
POST my-index-000001/_update/mxImkYkBbTg-D2mdeNsk
{
  "script": {
    "source": "if (ctx._source.workspaces !== null && !ctx._source.workspaces?.contains('public')) { ctx._source?.workspaces.add(params.workspace) }",
    "lang": "painless",
    "params": {
      "workspace": "test"
    }
  }
}
```

something like this.

Reply:
make sense, i will try convert to painless script